src/parsers/expression/selection.js

```diff
@@ -135,13 +161,33 @@ export function vlPointDomain(name, encoding, field, op) {
     }
   }
 
-  values = entries.reduce(function(acc, entry) {
-    acc.push({
-      unit: entry.unit,
-      value: entry.values[index]
-    });
-    return acc;
-  }, []);
+  // multi selections union within the same unit and intersect across units.
+  // if we've got only one unit, enforce union for more efficient materialization.
+  if (unitIdx && unitIdx.size === 1) {
+    op = UNION;
+  }
+
+  if (unitIdx && op === INTERSECT) {
+    units = entries.reduce(function(acc, entry) {
+      var u = acc[entry.unit] || (acc[entry.unit] = []);
+      u.push({unit: entry.unit, value: entry.values[index]});
+      return acc;
+    }, {});
+
+    values = Object.keys(units).reduce(function(acc, unit) {
+       acc.push({
+        unit: unit,
+        value: continuous ? continuousDomain(units[unit], UNION) :
+          discreteDomain(units[unit], UNION)
+      });
+      return acc;
+    }, []);
+  } else {
+    values = entries.reduce(function(acc, entry) {
+      acc.push({unit: entry.unit, value: entry.values[index]});
+      return acc;
+    }, []);
```

Comment:
Again: Why use `reduce` here? It looks like this could be done more parsimoniously with `map`.